### ISSUE TITLE
SystemCleaner not deleting empty directories in ../storage/emulated/0/Android/data

### ISSUE THREAD

**Issue by Gitoffthelawn** [d4rken/sdmaid-public#4532]
Hello Matthias!

I noticed that SystemCleaner is not deleting empty directories (aka folders) in in `../storage/emulated/0/Android/data`.

Is this intentional, or a bug?

If it's intentional, what is the reason, if you don't mind me asking?  Also, is there any harm in manually deleting empty directories in that location?

When performing manual backups, I go through all those folders to see which have data that needs to be backed up.  It's much faster and easier when I don't have to go though scores of empty folders.

Thanks!

**Comment by Sridhar-Ananthanarayanan** [comment]
Many apps recreate empty directories as soon as you delete them, or when you run those apps. It is also possible that the folders are not really empty, I mean they may contain hidden files.

Also, SD Maid requires multiple runs if there are multiple empty sub-folders.

If you are backing up app data folders, you should be backing up the parent folder as it is, ignoring empty sub-folders it may contain.

**Comment by Gitoffthelawn** [comment]
@TheMysticS wrote:
> Many apps recreate empty directories as soon as you delete them, or when you run those apps. It is also possible that the folders are not really empty, I mean they may contain hidden files.

Yes, some definitely do, and some don't.  SD Maid's SystemCleaner doesn't delete any otherwise-empty folders with hidden files does it?

> Also, SD Maid requires multiple runs if there are multiple empty sub-folders.

Correct.  I think I filed an issue about this a few years ago.   Currently, I run SystemCleaner several times until it reports that there is nothing left to clean.

> If you are backing up app data folders, you should be backing up the parent folder as it is, ignoring empty sub-folders it may contain.

I perform the backups by hand, because there are so few I want to actually backup.  So every once in a while I go through a the folders in `../storage/emulated/0/Android/data` and see if any are worth backing up.  If there were no empty folders in there, that process would be faster and more efficient.

**Comment by d4rken** [comment]
An folder with "hidden" files inside, is not empty, I'm unaware of any issues that suggest otherwise.

`Android/data` is excluded from the default empty directories filter by default as the system keeps recreating many empty folders there automatically. This means most deletions are in vain, and users are confused if directories keep reappearing after repeated attempts to delete them.
While the benefit is not clear, it should be possible to create a custom filter without this restriction that deletes empty directories.

> Also, SD Maid requires multiple runs if there are multiple empty sub-folders.

That's an unfortunate but known restriction of the SystemCleaner tool.

**Comment by Gitoffthelawn** [comment]
> A folder with "hidden" files inside, is not empty, I'm unaware of any issues that suggest otherwise.

Good to hear!  For a moment, I thought perhaps you were suggesting otherwise. :)

> `Android/data` is excluded from the default empty directories filter as the system keeps recreating many empty folders there automatically. This means most deletions are in vain, and users are confused if directories keep reappearing after repeated attempts to delete them.

I wasn't sure about this for myself, so I ran a test.  I had 279 empty folders in `Android/data`.  I deleted all of them.  Only 14 of them reappeared after a day's use.  So, it looks like a minority come back within a short time, and the rest either never come back, or more likely, only come back if certain conditions are met.

Interestingly, I then ran a bunch of apps that had empty folders before I performed the big purge.  I expected most to recreate their empty folders, but they didn't.  Only about a third of them recreated their folders.

> It should be possible to create a custom filter without this restriction that deletes empty directories.

Where were you a week ago?!? :) :) :) LOL! :)

I finally figured that out myself after doing this manually for who knows how long.  I had actually never created a custom filter before, so it was a good learning experience.  You've really made the custom filters very flexible and configurable.  Great job!

> > Also, SD Maid requires multiple runs if there are multiple empty sub-folders.
> 
> That's an unfortunate but known restriction of the SystemCleaner tool.

And I know the nice guy who can fix it!  <points to you :)>
